src/runtime-rs/crates/hypervisor/src/device/device_manager.rs

```diff
@@ -0,0 +1,304 @@
+// Copyright (c) 2019-2022 Alibaba Cloud
+// Copyright (c) 2019-2022 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use crate::{
+    device::blk_dev_manager::BlockDeviceManager,
+    device_type::{Device, DeviceArgument, GenericConfig},
+    DeviceManagerInner, DeviceType, Hypervisor,
+};
+use anyhow::{anyhow, Context, Result};
+use ini::Ini;
+use std::{collections::HashMap, sync::Arc};
+use tokio::sync::{Mutex, RwLock};
+
+use super::vfio_dev_manager::VfioDeviceManager;
+use super::vhost_dev_manager::VhostUserDeviceManager;
+pub type ArcBoxDevice = Arc<Mutex<Box<dyn Device>>>;
+const SYS_DEV_PREFIX: &str = "/sys/dev";
+
+/// VIRTIO_BLOCK_MMIO indicates block driver is virtio-mmio based
+pub const VIRTIO_BLOCK_MMIO: &str = "virtio-blk-mmio";
+/// VIRTIO_BLOCK_PCI indicates block driver is virtio-pci based
+pub const VIRTIO_BLOCK_PCI: &str = "virtio-blk-pci";
+pub const VFIO: &str = "vfio";
+pub const KATA_MMIO_BLK_DEV_TYPE: &str = "mmioblk";
+pub const KATA_BLK_DEV_TYPE: &str = "blk";
+use crate::device::AgentDevice;
+#[derive(Clone)]
+pub struct DeviceManager {
+    pub dev_managers: HashMap<DeviceType, Arc<RwLock<Box<dyn DeviceManagerInner + Send + Sync>>>>,
+    hypervisor: Arc<dyn Hypervisor>,
+    block_index: u64,
```

Comment:
Not get where you use block index and why does it needed in the `DeviceManager` instead of `BlockDeviceManager`

Reply:
`block_index` like is a global variable for the hypervisor block device. So `block_index` will be used by both VfioDeviceManager and `BlockDeviceManager`

Reply:
I think the `block_index` and `released_block_index` should maintain in `block_manager`.

Reply:
I don't think so, if we want to use vfio to attach the block device, the `block_index` cannot be fetched in `VfioDeviceManager`